### ISSUE TITLE
Cannot get include directories from cmake when `target_include_directories` is used

### ISSUE THREAD

**Issue by smfelsher** [niosus/EasyClangComplete#322]
# Please do this before submitting: #
- [X] check for already solved issues
- [X] try restarting Sublime Text

# Please state the following: #
- Sublime Text version: `3.0, Build 3143`
- which system are you on: `Ubuntu 16.04 64-bit`

clang version: 3.8.0-2ubuntu4

Here is the project I created to demonstrate my issue: [smfelsher/ecc_issue_demo ](https://github.com/smfelsher/ecc_issue_demo)

I have an application and a library defined in my project. Everything is being built successfully with CMake, so there are CMakeLists.txt files for each project and a top-level project.

In hello_world/Main.cpp, ECC is indicating that it can't find the header file Print.hpp. 

Now, when I set   

```
"verbose" : true,
"use_libclang" : false,
"show_type_info": false,
```
The ECC error markers go away, but the clang build command does not have the 'lib/include' in the include path

```
['clang++', '-c', '-fsyntax-only', '-x', 'c++', '-std=c++11', '-I/usr/include', '-I/usr/lib/clang/3.8.0/include', '-O0', '-g3', '-pedantic-errors', '-Wall', '-Wextra', '-Wdouble-promotion', '-Wmissing-include-dirs', '-Wunused', '-Wshadow', '-Wconversion', '-Wmissing-declarations', '-Werror', '/tmp/EasyClangComplete/Main.cpp']
```
Here are my issues:

1. Why did the error markers go away just by switching between use_libclang true/false?
1. Why is ECC not picking up the library include path from the libPrint/CMakeLists.txt `target_incude_directories`?
1. Changing options for ECC is sometimes causing SublimeText to freeze, which requires the application to be killed/"Force Quit".

**Comment by smfelsher** [comment]
Attached are two (2) verbose logs. My custom ECC settings were:
```
{
  "verbose" : true,
  "include_file_folder" : false,
  "include_file_parent_folder" : false,
  "common_flags" :
  [
    "-I/usr/include",
    "-I/usr/lib/clang/$clang_version/include",
  ]
}
```
[verbose_logs_use_lib_clang_true.txt](https://github.com/niosus/EasyClangComplete/files/1333334/verbose_logs_use_lib_clang_true.txt)

```
{
  "verbose" : true,
  "use_libclang" : false,
  "show_type_info": false,
  "include_file_folder" : false,
  "include_file_parent_folder" : false,
   "common_flags" :
  [
    "-I/usr/include",
    "-I/usr/lib/clang/$clang_version/include",
  ]
}
```
[verbose_logs_use_lib_clang_false.txt](https://github.com/niosus/EasyClangComplete/files/1333335/verbose_logs_use_lib_clang_false.txt)

**Comment by niosus** [comment]
I've looked in the first log and nothing there is wrong. It looks all ok to me. It has found a correct libclang and seems to parse and reparse the translation unit well.  The includes from cmake are still missing.
It is probably due to using `target_include_directories` as opposed to `include_directories`. This probably is due to cmake generating a compilation database of a style that we don't know how to parse. I will look into this. Thanks for all the infos. I guess now I have everything to reproduce this.

**Comment by smfelsher** [comment]
The `target_include_directories`, rather than `include_directories`, comes from a "best practices" document I read while learning CMake. Please let me know if you need any other information or assistance from me. I will do my best to help. I greatly appreciate the time and effort you have sacrificed for this tool.

**Comment by niosus** [comment]
Sure, I do not suggest you change it. I was not aware of it and it seems like a thing I should know of :) I am pretty sure it generates a slightly different database that the plugin does not know how to parse. Thanks again for this issue.

**Comment by niosus** [comment]
Sorry for a huge delay. It gets hard to find time for the plugin lately :( 
I've looked at this issue now. It seems to me that it is not an issue of a plugin. It seems that the project is configured in a way that I do not anticipate. To be more precise, there are a couple of things that seem odd to me:
1. Library is included *after* hello_world folder. I'm not sure how CMake can deal with it as it should be just reading the file sequentially.
2. Library is a separate project and hello_world folder is too. The plugin searches for the first CMakeLists.txt file that contains keyword "project" in it and tries to build it independently. In your case, I do not understand how CMake should know about the existence of the library in the `hello_world` project. In my mind, every CMake project should be completely independent from everything else and should have the means to find all the needed variables for all libraries it relies on. This assumption of mine is violated here.

I will close this issue for now. Feel free to continue discussing it here. I might be wrong in the way I treat it, but I strongly believe that the core issue here is that the `hello_world` project is not independent.

**Comment by smfelsher** [comment]
1. When you say "library is included _after_ hello_word" are you referring to my top-level CMakeLists.txt?
1. Doesn't CMake know about the library when it parses my top-level CMakeLists.txt file? 

I agree I may be violating the principle that each project should be able to be built completely independently. I will research CMake some more.

**Comment by smfelsher** [comment]
I have been experimenting with my [smfelsher/ecc_issue_demo](https://github.com/smfelsher/ecc_issue_demo) project. I can configure the library project to build independently. However, the hello_world project is dependent upon the library, so it needs the library project to be built before it will build. This is where the ECC plugin runs into an issue.

I am building the library, but not "installing" it into a fixed location. However, I do know that its out-of-source build is "../libPrint", so I can configure my CMakeLists.txt for hello_world to find that location. I see that the ECC plugin creates two independent CMake build directories in my "/tmp" directory. The /tmp/hello_world CMake project can't find the library path, so it doesn't define the CMake cache variable that points to the library's CMake config.

I think I may be able to overcome this issue by having CMake generate the JSON build commands and configure ECC to use that instead.

**Comment by niosus** [comment]
Ok, I've spent some more time on it and have good news. When I was designing the plugin I have added an option `search_in` for every flags source, one of which is CMakeLists.txt. See the PR I have added to your test project that configures the sublime text project in such a way, that it finds the top-level CMakeLists.txt, thus solving all your problems. Feel free to ask me anything if you still have issues. And sorry for the absence of good documentation.
